compiler/rustc_hir_typeck/src/fn_ctxt/mod.rs

```diff
@@ -279,19 +278,39 @@ impl<'a, 'tcx> AstConv<'tcx> for FnCtxt<'a, 'tcx> {
         self.tcx().mk_projection(item_def_id, item_substs)
     }
 
-    fn normalize_ty(&self, span: Span, ty: Ty<'tcx>) -> Ty<'tcx> {
-        if ty.has_escaping_bound_vars() {
-            ty // FIXME: normalization and escaping regions
-        } else {
-            self.normalize_associated_types_in(span, ty)
+    fn probe_adt(&self, span: Span, ty: Ty<'tcx>) -> Option<ty::AdtDef<'tcx>> {
+        match ty.kind() {
+            ty::Adt(adt_def, _) => Some(*adt_def),
+            // FIXME: We should normalize even with escaping bound vars.
```

Comment:
no, we shouldn't imo

imo we should have never relied on normalization during ast lowering. We won't be able to remove the existing normalization due to backwards compatibility. I think we should just live with the existing inconsistency, potentially even linting when normalization is required to remove it 2030 or whatever

Reply:
I tend to agree. The fact that path resolution depends on normalization may be an unintended behavior.

Reply:
cc @rust-lang/types I generally don't think we should normalize during astconv to resolve stuff like `Self::A` if `Self` is a projection. Examples include
```rust
trait Identity {
    type Id;
}
impl<T> Identity for T {
    type Id = T;
}

enum Foo {
    A,
    B,
}

trait Trait {
    fn return_a() -> Self;
}

impl Trait for <Foo as Identity>::Id {
    fn return_a() -> Self {
        Self::A
    }
}

fn main() {
    let x = <<Foo as Identity>::Id>::A;
}
```
We do have to keep supporting this for backwards compatibility however.

I therefore think that we should 1) remove the FIXME here and add a comment for "we cannot normalize inside of a binder" 2) rename this method to `probe_adt_normalize_for_backcompat` to prevent us from at least adding further uses of that method.

Reply:
This looks like T-lang issue. Can we open a new issue and nominate it for t-lang?

Anyway  I toned down the FIXME comment and opened a new ticket for it #104767. See also #103640.